incremental-font-transfer/src/uri_templates.rs

```diff
@@ -0,0 +1,682 @@
+                self.append_str(id_value);
+                Ok(ParseState::Literal)
+            }
+            (Variable::ID64, b'}') => {
+                // TODO percent encode any characters in id64 as needed.
+                self.append_str(id64_value);
+                Ok(ParseState::Literal)
+            }
+            (Variable::DX(digit), b'}') => {
+                self.append_id_digit(id_value, digit);
+                Ok(ParseState::Literal)
+            }
+            (Variable::Undefined, b'}') => {
+                // Undefined variable name just ignore it.
+                Ok(ParseState::Literal)
+            }
+            (Variable::PercentEncoding(_), b'}') => Err(UriTemplateError), // Unterminated percent encoding
+
+            // ### Percent Encoding Validation ###
+            (Variable::PercentEncoding(digit), _) => {
+                if !byte_info.is_hex_digit {
+                    return Err(UriTemplateError);
+                }
+
+                match digit {
+                    Digit::One => Ok(ParseState::Expression(Variable::PercentEncoding(
+                        Digit::Two,
+                    ))),
+                    Digit::Two => Ok(ParseState::Expression(Variable::Undefined)),
+                }
+            }
+
+            // ### Dot validity checking ###
+            (Variable::Begin, b'.') => Err(UriTemplateError), // . operator not allowed.
+            (Variable::Dot, b'}') | (Variable::Dot, b'.') => {
+                Err(UriTemplateError) // trailing . or .. is not allowed.
+            }
+            (_, b'.') => Ok(ParseState::Expression(Variable::Dot)),
+
+            // ### Enter percent encoding ###
+            (_, b'%') => Ok(ParseState::Expression(Variable::PercentEncoding(
+                Digit::One,
+            ))),
+
+            // ### Everything else ###
+            (_, b'}') => Err(UriTemplateError), // Unexpected closing brace
+
+            // Just skipping through an undefined variable name.
+            _ => Ok(ParseState::Expression(Variable::Undefined)),
+        }
+    }
+
+    /// Appends the expanded value of d1, d2, d3, or d4.
+    ///
+    /// See: <https://w3c.github.io/IFT/Overview.html#uri-templates>
+    fn append_id_digit(&mut self, id_value: &str, digit: u8) {
+        self.append(
+            *id_value
+                .len()
+                .checked_sub(digit.into())
+                .and_then(|index| id_value.as_bytes().get(index))
+                .unwrap_or(&b'_'),
+        )
+    }
+
+    // Appends a string to the output.
+    fn append_str(&mut self, value: &str) {
+        self.0.push_str(value)
+    }
+
+    // Appends a single byte to the output.
+    fn append(&mut self, byte: u8) {
+        self.0.push(byte.into());
+    }
+
+    // Appends the percent encoded representation (%XX) of a byte to the output.
+    fn append_percent_encoded(&mut self, byte: u8) {
+        self.0.push_str(&format!("%{:02X}", byte));
```

Comment:
`format!` allocates; you can skip that by writing directly into the buffer, like:

```rust
write!(&mut self.0, "%{:02X}", byte);
```